ghidra: documentation (#1759)
mandiant/capa: capa/ghidra/README.md

<div align="center">
    <img src="/doc/img/ghidra_backend_logo.png" width=300 height=175>
</div>

> The Ghidra feature extractor is an application of the FLARE team's open-source project, Ghidrathon, to integrate capa with Ghidra. capa is a framework that uses a well-defined collection of rules to identify capabilities in a program. You can run capa against a PE file, ELF file, or shellcode and it tells you what it thinks the program can do. For example, it might suggest that the program is a backdoor, can install services, or relies on HTTP to communicate. The Ghidra feature extractor can be used to run capa analysis on your Ghidra databases without needing access to the original binary file.

Reviewer: ```suggestion
The Ghidra feature extractor is an application of the FLARE team's open-source project, Ghidrathon, to integrate capa with Ghidra using Python 3. capa is a framework that uses a well-defined collection of rules to identify capabilities in a program. You can run capa against a PE file, ELF file, or shellcode and it tells you what it thinks the program can do. For example, it might suggest that the program is a backdoor, can install services, or relies on HTTP to communicate. The Ghidra feature extractor can be used to run capa analysis on your Ghidra databases without needing access to the original binary file.
```

Reviewer: Let's also be explicit here that this is a Python 3 project.

## Getting Started

### Installation

| Dependency | Version | Source |
|------------|---------|--------|
| Ghidrathon | `>= 3.0.0` | https://github.com/mandiant/Ghidrathon |

In order to run capa using the Ghidra feature extractor, you must install capa as a library and obtain the offical capa rules that match the version you have installed. The following steps may be followed:

1. Install capa and its dependencies from PyPI using the Python interpreter configured for your Ghidra installation:
```bash
$ pip install flare-capa
```

2. Download and extract the [official capa rules](https://github.com/mandiant/capa-rules/releases) that match the version you have installed
   1. Use the following command to view the version of capa you have installed:
```bash
$ pip show flare-capa
OR
$ capa --version
```

> 3. Copy `capa_ghidra.py` to your Ghidra user scripts directory OR update the `Ghidra Script Manager` path to point to it.

Reviewer: I've found it best to place important links as close to the first time you mention them. Here, we provide the link to `capa_ghidra.py` so a user has immediate access.

   1. This entrypoint script is located in `capa_install_dir/capa/ghidra/`

Once Ghidrathon is configured, you may now invoke capa from within Ghidra in three different ways. Each method suits different use cases of capa, and they include Ghidra's `headlessAnalyzer`, `Scripting Console`, and `Script Manger`.

## Running capa with the Ghidra feature extractor

> ### Ghidra's headlessAnalyzer

Reviewer: ```suggestion
### Ghidra's Headless Analyzer
```

> To invoke capa headlessly (i.e. without the Ghidra user interface), we must call the `analyzeHeadless` script provided in your `$GHIDRA_INSTALL_DIR/support` and point it towards capa's `capa_ghidra.py`. One thing to note is that capa runs as a `PostScript`, as in post-analysis script, so we need to provide `analyzeHeadless` with the path and script to run against our project. The preferred method for the Ghidra feature extractor is the entrypoint script, `/capa/ghidra/capa_ghidra.py`. Additional capa command line arguments must be provided in a single, space-delimited string i.e. `"/path/to/rules -v"`. To display the help & usage statement, the keyword `help` must be used instead of the typical `-h or --help`.

Reviewer: This many moving parts is a hard one to explain in a single paragraph. Let's focus on the details that are most relevant to the user.

The syntax is as so:
```bash

> ./$GHIDRA_INSTALL_DIR/support/analyzeHeadless /path/to/gpr_dir/ gpr_name -process sample_name.exe_ -ScriptPath /path/to/capa_install/capa/ghidra -PostScript capa_ghidra.py "/path/to/rules/"

Reviewer: ```suggestion
./$GHIDRA_INSTALL_DIR/support/analyzeHeadless /path/to/gpr_dir/ gpr_name -process sample_name.exe_ -ScriptPath /path/to/capa_ghidra.py/ -PostScript capa_ghidra.py "/path/to/capa/rules/"
```

Reviewer: With the introduction of `capa_ghidra.py`, we no longer require users know the full path to their capa installation, just that they know the path to their local `capa_ghidra.py`.

```

> > **Note:** You may add the `$GHIDRA_INSTALL_DIR/support` to your `$PATH` in order to call `analyzeHeadless` as a standalone program.

Reviewer: Let's add additional context for the `-ScriptPath` argument to help users better understand how they can configure their environment to make these commands more streamlined.

> If you do not have an existing ghidra project, you may also create one with the headlessAnalyzer via the `-Import` flag. Post scripts may also be ran in the same invocation.

Reviewer: Please correct me here but my understanding is that `-Import` can only be used to import a new sample into an _existing_ Ghidra project, but not be used to create a new Ghidra project?

> The syntax to both import a new file and run capa against it is:

Reviewer: Stated above now.

```bash

> ./$GHIDRA_INSTALL_DIR/support/analyzeHeadless /path/to/gpr_dir/ gpr_name -Import /path/to/sample_name.exe_ -ScriptPath /path/to/capa_install/capa/ghidra -PostScript capa_ghidra.py "/path/to/rules/"

Reviewer: Let's be more explicit about what where saying here, e.g. by removing `.exe_` we don't artificially restrict users into thinking this integration only supports PE exe files.

```

> > **Note:** The `/path/to/gpr_dir/` must exist before importing a new project into it.

Reviewer: Please see my comment above about the Ghidra projects and the `-Import` flag.

mbc        Process::Check Mutex [C0043]
basic block @ 0x1000102E in function 0x10001010
  and:
    or:
      api: kernel32.OpenMutex @ 0x10001059

create mutex
namespace  host-interaction/mutex
author     moritz.raabe@mandiant.com, michael.hunhoff@mandiant.com
scope      function
mbc        Process::Create Mutex [C0042]
function @ 0x10001010
  or:
    api: kernel32.CreateMutex @ 0x1000106E

create process on Windows
namespace  host-interaction/process/create
author     moritz.raabe@mandiant.com
scope      basic block
mbc        Process::Create Process [C0017]
basic block @ 0x10001179 in function 0x10001010
  or:
    api: kernel32.CreateProcess @ 0x100011AF

Script /ghidra_scripts/capa_ghidra.py called exit with code 0
INFO  ANALYZING changes made by post scripts: /Practical Malware Analysis Lab 01-01.dll_ (HeadlessAnalyzer)  

[...]
```

### Ghidra's Script Manager

To invoke capa from the `Ghidra Script Manager`, open your Ghidra Project's Code Browser and open the `Script Manager` window by navigating to `Window -> Script Manager`. Select `capa_ghidra.py` and run the script. capa will then prompt you to choose a `rules` directory and specify the output verbosity level. 

> > **Note:** In order for the Script Manager to recognize `capa_ghidra.py` you must either copy it to your Ghidra user scripts directory or update the Script Manager search path to include directory that contains it.

Reviewer: We should always be explicit about what a user should expect before, during, and after using our tool (even though we show this later using screen captures). Here I expand on what you had already provided to include what information they will be asked to provide during script execution and where the results go when the script is finished.

> Successful invocation from the `Ghidra Script Manager` will look like:

Reviewer: ```suggestion
```

Reviewer: Covered above now.

<div align="center">
    <img src="/doc/img/ghidra_script_mngr_rules.png">
    <img src="/doc/img/ghidra_script_mngr_verbosity.png">
    <img src="/doc/img/ghidra_script_mngr_output.png">
</div>

### Ghidrathon's Script Console

To invoke capa from Ghidrathon's Script Console, open your Ghidra project's Code Browser and open the `Ghidrathon` window by navigating to `Window -> Ghidrathon`.

You must import capa into the console and run it via:

> 

Reviewer: Stripping out some information we can assume the user already knows, e.g. opening the code browser.

<div align="center">
    <img src="/doc/img/ghidra_backend_logo.png" width=300 height=175>
</div>

The Ghidra Backend is an application of the FLARE team's open-source project, Ghidrathon, to integrate capa with Ghidra. capa is a framework that is able to extract features from binaries and use that data to match against a curated collection of well-defined rules to identify capabilities in a program. You can use capa to run against PE files, ELF files, or shellcode and it tells you what it thinks the program can do. For example, it might suggest that the program is a backdoor, can install services, or relies on HTTP to communicate. The Ghidra Backend will will run capa analysis on the databases present in your Ghidra projects (.gpr) without needing access to the original binary file. Once a project has been analyzed, the Ghidra Backend helps you identify interesting areas of a program and build new capa rules using features extracted from your Ghidra projects.

## Installation & Usage:

> **Required Dependencies:**

Reviewer: ```suggestion
**Dependencies:**
```

| Dependency | Version | Source |
|------------|---------|--------|

> | capa | `>= 6.1.0` | `<url for capa release>`

Reviewer: ```suggestion
```

Reviewer: User's should already expect to have capa installed by this point.

Reviewer: We need to add directions for how a user must install capa in order to use it from Ghidra e.g. see the installation steps [specified for the IDA plugin.](https://github.com/mandiant/capa/tree/master/capa/ida/plugin#installation)

Namely, a user must:

1. Install capa as a library (see IDA plugin for command)
2. Download and extract official capa rules that matches the version of capa installed
3. Make it work with Ghidra from there

> | Ghidrathon | `>= 3.0.0` | https://github.com/mandiant/Ghidrathon/tree/v3.0.0 |

Reviewer: ```suggestion
| Ghidrathon | `>= 3.0.0` | https://github.com/mandiant/Ghidrathon |
```

Reviewer: With this change we do not need to update the link with each new Ghidrathon release.

> **note:** Please follow the Ghidrathon installation guide to ensure all additional dependency requirements are met and that it is properly configured to work with Ghidra

> Once Ghidrathon is configured and ready-to-go, you may now invoke capa from within Ghidra in three different ways. Each method suits different use cases of capa, and they include the `headlessAnalyzer`, `Scripting Console`, and `Script Manger`.

Reviewer: ```suggestion
Once Ghidrathon is configured, you may now invoke capa from within Ghidra in three different ways. Each method suits different use cases of capa, and they include the Ghidra's `headlessAnalyzer`, `Scripting Console`, and `Script Manger`.
```

> ### headlessAnalyzer

Reviewer: ```suggestion
### Ghidra's headlessAnalyzer
```

To invoke capa headlessly (i.e. without the Ghidra user interface), we must call the `analyzeHeadless` script provided in your `$GHIDRA_INSTALL_DIR` and point it towards capa's `main.py`. One thing to note is that capa runs as a `PostScript`, as in post-analysis script, so we need to provide `analyzeHeadless` with the path and script to run against our project.

The syntax is as so:
```bash
./$GHIDRA_INSTALL_DIR/support/analyzeHeadless /path/to/gpr_dir/ gpr_name -process sample_name.exe_ -ScriptPath /path/to/capa_install/capa -PostScript main.py
```

> > **note:** You may add the `$GHIDRA_INSTALL_DIR/support` to your `$PATH` in order to call `analyzeHeadless` as a standalone program.

Reviewer: ```suggestion
> **Note:** You may add the `$GHIDRA_INSTALL_DIR/support` to your `$PATH` in order to call `analyzeHeadless` as a standalone program.
```

If you do not have an existing ghidra project, you may also create one with the headlessAnalyzer via the `-Import` flag. Post scripts may also be ran in the same invocation.

The syntax to both import a new file and run capa against it is:
```bash
./$GHIDRA_INSTALL_DIR/support/analyzeHeadless /path/to/gpr_dir/ gpr_name -Import /path/to/sample_name.exe_ -ScriptPath /path/to/capa_install/capa -PostScript main.py
```
> **note:** The `/path/to/gpr_dir/` must exist before importing a new project into it.

A successful headlessAnalyzer run should look like:

> <img src="/doc/img/ghidrathon_headless.gif">

Reviewer: Let's stick with static screenshots if we can, to avoid the large size of these GIFs. GIFs are also difficult to recreate when changes are made to the project.
